### PR TITLE
bootchart: Account CPU time spent in non-main threads of the process (v3)

### PR DESCRIPTION
**Pull request by gmacario**
Hello @teg,

I have implemented the remaining changes as you suggested.

Here is v3 of the fix to Bug https://github.com/systemd/systemd/issues/139

Thanks for your patience,
Gianpaolo

### CODE REVIEW

**Comment by zonque** [comment]
Please squash the two commits in this PR into one.

File: src/bootchart/store.c [hunk 1]
```diff
@@ -409,6 +410,61 @@ int log_sample(DIR *proc,
...
 
+                /* Take into account CPU runtime/waittime spent in non-main threads of the process
+                 * by parsing "/proc/[pid]/task/[tid]/schedstat" for all [tid] != [pid]
+                 * See https://github.com/systemd/systemd/issues/139
+                 */
+
+                /* Browse directory "/proc/[pid]/task" to know the thread ids of process [pid] */
+                snprintf(filename, sizeof(filename), PID_FMT "/task", pid);
+                taskfd = openat(procfd, filename, O_RDONLY|O_DIRECTORY|O_CLOEXEC);
+                if (taskfd != -1) {
```
Comment (zonque) [inline src/bootchart/store.c:421]: While at it, turn this into `>= 0` to put it in line with other code.

File: src/bootchart/store.c [hunk 2]
```diff
@@ -409,6 +410,61 @@ int log_sample(DIR *proc,
...
+                        _cleanup_closedir_ DIR *taskdir = NULL;
+
+                        taskdir = fdopendir(taskfd);
+                        FOREACH_DIRENT(ent, taskdir, goto task_next) {
```
Comment (zonque) [inline src/bootchart/store.c:425]: Again, why not use `break` as the last argument here? You'd need to remove the `_cleanup_close_` attribute from `int taskfd`, which isn't needed anyway AFAICS …

**Comment by gmacario** [comment]
Superseded by PR https://github.com/systemd/systemd/pull/365

**Pull request closed without merging**